source/common/http/filter/fault_filter.cc

```diff
@@ -0,0 +1,120 @@
+#include "fault_filter.h"
+
+#include "envoy/event/timer.h"
+#include "envoy/http/codes.h"
+#include "envoy/http/header_map.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/stats/stats.h"
+
+#include "common/common/assert.h"
+#include "common/common/empty_string.h"
+#include "common/common/enum_to_int.h"
+#include "common/http/codes.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+
+namespace Http {
+FaultFilter::FaultFilter(FaultFilterConfigPtr config) : config_(config) {}
+
+FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
+
+/**
+ * Delays and aborts are independent events. One can inject a delay
+ * followed by an abort or inject just a delay or abort. In this callback,
+ * if we inject a delay, then we will inject the abort in the delay timer
+ * callback.
+ */
+FilterHeadersStatus FaultFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // Check for header matches first
+  if (!matches(headers)) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (config_->runtime_.snapshot().featureEnabled("fault.http.delay_enabled",
+                                                  config_->delay_enabled_)) {
+    delay_timer_ = callbacks_->dispatcher().createTimer([this]() -> void { postDelayInjection(); });
+    delay_timer_->enableTimer(std::chrono::milliseconds(config_->runtime_.snapshot().getInteger(
+        "fault.http.delay_duration", config_->delay_duration_)));
+    config_->stats_.delays_injected_.inc();
+    return FilterHeadersStatus::StopIteration;
+  } else if (config_->runtime_.snapshot().featureEnabled("fault.http.abort_enabled",
+                                                         config_->abort_enabled_)) {
+    Http::HeaderMapPtr response_headers{new HeaderMapImpl{
+        {Headers::get().Status, std::to_string(enumToInt(config_->runtime_.snapshot().getInteger(
+                                    "fault.http.abort_code", config_->abort_code_)))}}};
+    callbacks_->encodeHeaders(std::move(response_headers), true);
+    config_->stats_.aborts_injected_.inc();
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  return FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus FaultFilter::decodeData(Buffer::Instance&, bool end_stream) {
+  if (end_stream) {
+    resetInternalState();
+  }
+  return FilterDataStatus::Continue;
+}
+
+FilterTrailersStatus FaultFilter::decodeTrailers(HeaderMap&) {
+  resetInternalState();
+  return FilterTrailersStatus::Continue;
+}
+
+FaultFilterStats FaultFilter::generateStats(const std::string& prefix, Stats::Store& store) {
+  std::string final_prefix = prefix + "fault.";
+  return {ALL_FAULT_FILTER_STATS(POOL_COUNTER_PREFIX(store, final_prefix))};
+}
+
+// header match semantics in fault filter works is same as the one in route block
+bool FaultFilter::matches(const Http::HeaderMap& headers) const {
```

Comment:
instead of copying this code from the route table, can we move the code into a static helper, perhaps in ConfigUtility class, and then call from both places, just passing in the headers and the headers list.